pkg/controller/deployment/failed.go

```diff
@@ -0,0 +1,168 @@
+		// was successfully deployed. If the condition already exists, ignore this update.
+		// Cleanup any condition that reports lack of progress.
+		if !util.DeploymentConditionExists(newStatus, extensions.DeploymentProgressing, api.ConditionTrue, util.NewRSAvailableReason) {
+			msg := fmt.Sprintf("Replica set %q has successfully progressed.", newRS.Name)
+			condition := util.NewDeploymentCondition(extensions.DeploymentProgressing, api.ConditionTrue, util.NewRSAvailableReason, msg)
+			util.SetDeploymentCondition(&newStatus, *condition)
+		}
+		// Cleanup conditions that denote any failures since it may be confusing for users.
+		util.RemoveDeploymentCondition(&newStatus, extensions.DeploymentReplicaFailure)
+
+	case util.IsDeploymentProgressing(d, &newStatus):
+		// If there is any progress made, continue by not checking if the deployment failed. This
+		// behavior emulates the rolling updater progressDeadline check.
+		msg := fmt.Sprintf("Replica set %q is progressing.", newRS.Name)
+		condition := util.NewDeploymentCondition(extensions.DeploymentProgressing, api.ConditionTrue, util.ReplicaSetUpdatedReason, msg)
+		util.SetDeploymentCondition(&newStatus, *condition)
+
+	case util.IsDeploymentFailed(d, &newStatus):
+		// Update the deployment with a timeout condition. If the condition already exists,
+		// ignore this update.
+		if !util.DeploymentConditionExists(newStatus, extensions.DeploymentProgressing, api.ConditionFalse, util.TimedOutReason) {
+			msg := fmt.Sprintf("Replica set %q has timed out progressing.", newRS.Name)
+			condition := util.NewDeploymentCondition(extensions.DeploymentProgressing, api.ConditionFalse, util.TimedOutReason, msg)
+			util.SetDeploymentCondition(&newStatus, *condition)
+		}
+	}
+
+	// Move warning events of the replica set in deployment conditions. Let's not display
+	// these warnings once a deployment completes, otherwise it may be confusing for users.
+	if !util.IsDeploymentComplete(d, &newStatus) {
+		for _, warning := range dc.getReplicaFailures(newRS) {
+			if !util.DeploymentConditionExists(newStatus, warning.Type, warning.Status, warning.Reason) {
+				util.SetDeploymentCondition(&newStatus, warning)
+			}
+		}
+	}
+
+	if reflect.DeepEqual(d.Status, newStatus) {
+		return nil
+	}
+
+	newDeployment := d
+	newDeployment.Status = newStatus
+	_, err = dc.client.Extensions().Deployments(newDeployment.Namespace).UpdateStatus(newDeployment)
+	return err
+}
+
+// getReplicaFailures transitions all the warning events found for a replica set into
+// deployment conditions. It's up to the caller to dedupe between the conditions and
+// add just a single condition, ideally the oldest one.
+// TODO: Once #32863 is fixed we should stop listing events and instead look into replica
+// set Conditions for failures.
+func (dc *DeploymentController) getReplicaFailures(rs *extensions.ReplicaSet) []extensions.DeploymentCondition {
+	eventList, err := dc.client.Core().Events(rs.Namespace).Search(rs)
```

Comment:
Every time we rollout a new template until the deployment completes plus it seems to be called for all failed deployments every time they are synced ie. 30s is the current interval for relisting. We can avoid relisting events by inspecting if the ReplicaFailure condition already exists in the deployment but we cannot be sure that something new didn't occur. I wouldn't care much if we get Conditions for ReplicaSets for 1.5.